promql/engine.go

```diff
@@ -1399,15 +1399,36 @@ func (ev *evaluator) vectorSelectorSingle(it *storage.BufferedSeriesIterator, no
 
 var pointPool = sync.Pool{}
 
-func getPointSlice(sz int) []Point {
+func getPointSlice(maxSz int) []Point {
 	p := pointPool.Get()
 	if p != nil {
 		return p.([]Point)
 	}
+
+	// Some queries can blow up the heap (and then run into maxSamples)
```

Comment:
Some queries however will use what they asked for, and giving them a smaller slice will only make them slower as it'll have to be extended many times.